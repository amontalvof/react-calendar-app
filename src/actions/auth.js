import Swal from 'sweetalert2';
import { fetchSinToken } from '../helpers/fetch';
import { types } from '../types/types';

export const startLogin = (email, password) => {
    return async (dispatch) => {
        const resp = await fetchSinToken('auth', { email, password }, 'POST');
        const body = await resp.json();
        if (body.ok) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-time', new Date().getTime());
            dispatch(
                login({
                    uid: body.uid,
                    name: body.name,
                })
            );
        } else {
            Swal.fire('Error!', body.message, 'error');
        }
    };
};

export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        const resp = await fetchSinToken(
            'auth/new',
            { email, password, name },
            'POST'
        );
        const body = await resp.json();
        if (body.ok) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-time', new Date().getTime());
            dispatch(
                login({
                    uid: body.uid,
                    name: body.name,
                })
            );
        } else {
            Swal.fire('Error!', body.message, 'error');
        }
    };
};

const login = (user) => ({
    type: types.authLogin,
    payload: user,
});
