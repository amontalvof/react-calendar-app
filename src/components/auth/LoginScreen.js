import React from 'react';
import { useDispatch } from 'react-redux';
import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const [formLoginValues, handleLoginInputChange] = useForm({
        loginEmail: 'fernando@gmail.com',
        loginPassword: '123456',
    });

    const { loginEmail, loginPassword } = formLoginValues;

    const handleLogin = (event) => {
        event.preventDefault();
        dispatch(startLogin(loginEmail, loginPassword));
    };

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Log in</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Email"
                                name="loginEmail"
                                value={loginEmail}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                name="loginPassword"
                                value={loginPassword}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Sign in</h3>
                    <form>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Name"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repeat password"
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Create account"
                            />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
};
