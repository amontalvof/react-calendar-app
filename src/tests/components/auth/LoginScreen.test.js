import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);

describe('Tests in <LoginScreen />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should be displayed correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should call the login dispatch', () => {
        wrapper.find('input[name="loginEmail"]').simulate('change', {
            target: {
                name: 'loginEmail',
                value: 'andy@gmail.com',
            },
        });

        wrapper.find('input[name="loginPassword"]').simulate('change', {
            target: {
                name: 'loginPassword',
                value: '123456',
            },
        });
        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault() {},
        });

        expect(startLogin).toHaveBeenCalledWith('andy@gmail.com', '123456');
    });

    test('should no sign in if passwords are different', () => {
        wrapper.find('input[name="registerPassword1"]').simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '123456',
            },
        });

        wrapper.find('input[name="registerPassword2"]').simulate('change', {
            target: {
                name: 'registerPassword2',
                value: '1234567',
            },
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault() {},
        });

        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledWith(
            'Error!',
            'Passwords must be the same.',
            'error'
        );
    });

    test('should sign in with equals passwords', () => {
        wrapper.find('input[name="registerName"]').simulate('change', {
            target: {
                name: 'registerName',
                value: 'Andy',
            },
        });

        wrapper.find('input[name="registerEmail"]').simulate('change', {
            target: {
                name: 'registerEmail',
                value: 'andy@gmail.com',
            },
        });

        wrapper.find('input[name="registerPassword1"]').simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '123456',
            },
        });

        wrapper.find('input[name="registerPassword2"]').simulate('change', {
            target: {
                name: 'registerPassword2',
                value: '123456',
            },
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault() {},
        });

        expect(Swal.fire).not.toHaveBeenCalled();
        expect(startRegister).toHaveBeenCalledWith(
            'andy@gmail.com',
            '123456',
            'Andy'
        );
    });
});
