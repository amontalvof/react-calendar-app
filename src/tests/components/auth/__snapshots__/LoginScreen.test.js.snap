// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> should be displayed correctly 1`] = `
<div
  className="login-container"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-6 login-form-1"
      >
        <h3>
          Log in
        </h3>
        <form
          onSubmit={[Function]}
        >
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="loginEmail"
              onChange={[Function]}
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="loginPassword"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
            <p
              className="helpTextLogin"
            >
              The password must be at least 6 characters long.
            </p>
          </div>
          <div
            className="form-group"
          >
            <input
              className="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        className="col-md-6 login-form-2"
      >
        <h3>
          Sign in
        </h3>
        <form
          onSubmit={[Function]}
        >
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="registerName"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="registerEmail"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="registerPassword1"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="registerPassword2"
              onChange={[Function]}
              placeholder="Repeat password"
              type="password"
              value=""
            />
            <p
              className="helpTextRgister"
            >
              The password must be at least 6 characters long.
            </p>
          </div>
          <div
            className="form-group"
          >
            <input
              className="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
